src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Moq;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Logging;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Utilities;
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.Consensus
+{
+    public class ChainedHeaderTreeTest
+    {
+        public class TestContext
+        {
+            public Network Network = Network.RegTest;
+            public Mock<IChainedHeaderValidator> ChainedHeaderValidatorMock = new Mock<IChainedHeaderValidator>();
+            public Mock<ICheckpoints> CheckpointsMock = new Mock<ICheckpoints>();
+            public Mock<IChainState> ChainStateMock = new Mock<IChainState>();
+            public ConsensusSettings ConsensusSettings = new ConsensusSettings();
+
+            public ChainedHeaderTree ChainedHeaderTree;
+
+            public ChainedHeaderTree CreateChainedHeaderTree()
+            {
+                this.ChainedHeaderTree = new ChainedHeaderTree(this.Network, new ExtendedLoggerFactory(), this.ChainedHeaderValidatorMock.Object, this.CheckpointsMock.Object, this.ChainStateMock.Object, this.ConsensusSettings);
+                return this.ChainedHeaderTree;
+            }
+
+            public ChainedHeader ExtendAChain(int count, ChainedHeader chainedHeader = null)
+            {
+                ChainedHeader previousHeader = chainedHeader ?? new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GenesisHash, 0);
+
+                foreach (int index in Enumerable.Range(1, count))
+                {
+                    BlockHeader header  = this.Network.Consensus.ConsensusFactory.CreateBlockHeader();
+                    header.HashPrevBlock = previousHeader.HashBlock;
+                    header.Bits = previousHeader.Header.Bits - 1000; // just increase difficulty.
+                    ChainedHeader newHeader = new ChainedHeader(header, header.GetHash(), previousHeader);
+                    previousHeader = newHeader;
+                }
+
+                return previousHeader;
+            }
+
+            public List<BlockHeader> ChainedHeaderToList(ChainedHeader chainedHeader, int count)
+            {
+                List<BlockHeader> list = new List<BlockHeader>();
+
+                ChainedHeader current = chainedHeader;
+                while (count > 0)
+                {
+                    list.Add(current.Header);
+                    current = current.Previous;
+                    count--;
+                }
+
+                list.Reverse();
```

Comment:
Hi, you're doing a reverse here.   So does this method assume you're at the tip and are producing a list going back in time?  

As an idea - If I supplied a minus value for count you go back in the chain and a plus value go forward?

Reply:
Ah we cant really go forward from a single ChainedHeader as we only have one prev but may have more then one Next.

I could though change the insert so always insert at position 0.
` list.Add(0, current.Header);`

Reply:
You could just preallocate the whole thing and insert to collect positions without doing a reverse. Definitely don't do insert the beginning of the list.